### PR TITLE
Port Nix-Perl to Meson

### PR DESCRIPTION
**Pull request by p01arst0rm**
# Motivation
The current buildsystem impacts Nix portability, and obstructs development of Nix on non GNU/Linux systems. Replacing the buildsystem with meson fixes this, along with adding better access to developer tools (i.e. valgrind).

# Context
https://github.com/NixOS/nix/pull/3160
https://github.com/NixOS/rfcs/pull/132
https://github.com/NixOS/nix/issues/9342
https://github.com/NixOS/nixpkgs/issues/26850

### CODE REVIEW

File: perl/default.nix
Comment (Ericson2314) [inline perl/default.nix:22]: This can become
```suggestion
    root = ./.;
```
now

File: perl/default.nix
Comment (Ericson2314) [inline perl/default.nix:24]:
```suggestion
```
No longer used since you pass in with the configure flags

File: perl/.yath.rc
Comment (Ericson2314) [inline perl/.yath.rc:3]: Probably shouldn't assume the build directory. We can pass in a `-I` to `yath` instead, leaving this file as is just for local dev.

File: perl/default.nix
Comment (Ericson2314) [inline perl/default.nix:36]:
```suggestion
    [
```
trailing space

**Comment by Ericson2314** [comment]
I think `]` and even `)` on their own line with trailing commas for the last list item / argument would be preferred.

File: perl/lib/Nix/meson.build
Comment (dcbaker) [inline perl/lib/Nix/meson.build:47]: Meson will inject this automatically in all cases

File: perl/meson_options.txt
Comment (dcbaker) [inline perl/meson_options.txt:11]: This is an unfortunate requirement, since it means that outside of configuration through the nix derivation the version will be unset. Meson can actually read a file formatted like the `.version` file is, and automatically set the version by adding `version: files('.version')` to the `project()` call, if you can symlink the `.version` file into the nix-perl directory

**Review by dcbaker — CHANGES_REQUESTED** [review]
There's an inconsistant use of spaces on both sides of `:`, I'm partial to the space on each side style, but probably should just pick one or the other and stick with it.

File: perl/meson_options.txt
Comment (p01arst0rm) [reply perl/meson_options.txt:11]: it would be good to have a meson builtin something like `meson setup build -Dmeson.project_version="1.2.3"`

Comment (dcbaker) [reply perl/meson_options.txt:11]: That doesn't seem useful to me. A project version is what it is, having a user arbitrarily change is asking for trouble.

I don't see what's wrong with a simple:
```console
cd perl
ln -s .version ../.version
git add .version
```
and adding `version : files('.version')` to the Meson.buidl

Comment (p01arst0rm) [reply perl/meson_options.txt:11]: symlinks aren't portable and can cause problems in different environments. the symlink would need to be created at buildtime. its messy, but it works just fine for now.

**Comment by Ericson2314** [comment]
> the style isnt standard across the codebase, there doesnt seem to be anything to suggest which should be used.

For C++ we put the `)` on the same line, but for `{ ... }` initializer lists we always put the `}` on its own line. That would be the equivalent of `[ ... ]`.

File: perl/meson_options.txt
Comment (Ericson2314) [inline perl/meson_options.txt:17]: Meson doesn't have an automatic version of this? (I forget)

File: perl/meson_options.txt
Comment (Ericson2314) [inline perl/meson_options.txt:35]: This is new to this build system, right? If so, I think we should probably skip for now and add later, since it is a "new feature".

File: perl/default.nix
Comment (Ericson2314) [inline perl/default.nix:61]: Check out `lib.mesonOption`
```suggestion
    (lib.mesonOption "version" (builtins.readFile ../.version))
    (lib.mesonOption "dbi_path" "${perlPackages.DBI}/${perl.libPrefix}")
    (lib.mesonOption "dbd_sqlite_path" "${perlPackages.DBDSQLite}/${perl.libPrefix}")
```

File: perl/meson_options.txt
Comment (dcbaker) [reply perl/meson_options.txt:17]: No, I’ve argued for one but other people don’t like the idea. Maybe I should just write some code

File: perl/meson_options.txt
Comment (p01arst0rm) [reply perl/meson_options.txt:35]: it isnt new, its already in the make files. this is like for like as close as possible

**Comment by roberth** [comment]
I love to see progress on this.

Just a small point: could you use the imperative mood instead of past tense in commit messages?
Nix, Nixpkgs and Git itself use it that way.

File: perl/meson_options.txt
Comment (dcbaker) [reply perl/meson_options.txt:35]: They're not behind an option though, they're just universally turned on. If the maintainers like this apporach that's fine, but it sure feels llike the `add_project_arguments()` approach I suggested initially would be a more 1:1 mapping of what the Makefiles had

File: perl/meson_options.txt
Comment (eli-schwartz) [reply perl/meson_options.txt:11]: Then use:
```meson
version: run_command('cat_versionfile.pl', '../version's, check: true).stdout().strip()
```

Or simply

```meson
version: files('../version')
```

Which I believe should work since meson doesn't care where the file is, just that it exists. If it's outside of the build directory and the project is refactored so the version file is not at the right relative location, that's a project issue, not a sandbox bug.

I really think you're overcomplicating this and there's no justification for meson to add a built-in option to erroneously ignore the upstream version number and break ABI for any project that actually uses the version number for something other than printing the meson summary().

File: perl/meson_options.txt
Comment (p01arst0rm) [reply perl/meson_options.txt]: switched to the static setup that is already in the build for now. maybe in the future this could cause issues if any of those flags needed overriding, but matching the current build is probably the best idea for porting;- improvements can build from that

File: perl/t/meson.build
Comment (Ericson2314) [inline perl/t/meson.build:9]: Let's use two spaces, and arrays, like https://mesonbuild.com/Style-guide.html suggests
```suggestion
nix_perl_tests = files([
  'init.t',
])
```
(and also trailing commas)

Comment (eli-schwartz) [reply perl/t/meson.build:9]: The style guide is wrong and should be fixed.

At minimum the use of arrays is redundant and produces noise. I accept that two spaces vs four is a bikeshed and while I sort of wish two spaces would not be pushed I may be in the minority here, but the array thing is absolutely definitely redundant.

Comment (p01arst0rm) [reply perl/t/meson.build:9]: a posssible alternative would be to create a project centric style guide or even a linter config

Comment (p01arst0rm) [reply perl/t/meson.build:9]: afaik `files()` returns an array object so it shouldnt be required to pass an array in, itll only get squashed anyway

Comment (eli-schwartz) [reply perl/t/meson.build:9]: `muon fmt` defaults to 4 spaces :D 

positional arguments to functions which take variadic arguments get flattened. Pass two arguments in, it gets interpreted as an array of posargs with two elements. Pass one array in, it gets interpreted as an array of posargs with one array-element in, and the array gets flattened to two elements.

Pass three arguments, two of which are arrays and one of which is a string, and it gets interpreted as an array of posargs with 3 elements, two of which get unpacked and merged together + the third standalone string. End result: a single array of strings that contains the union of strings reachable in the posargs.

This transparently handles `files('foo.c', core_sources, extra_sources)`.

Comment (eli-schwartz) [reply perl/t/meson.build:9]: Yes, that looks like what I'd do.

**Comment by roberth** [comment]
@Ericson2314 is this stuck on style alone?

**Comment by Ericson2314** [comment]
@roberth Yes, if that is even a blocker.

**Comment by p01arst0rm** [comment]
> @Ericson2314 is this stuck on style alone?

style can be fixed trivially, waiting on consensus of whether to merge before investing more time :)

**Comment by p01arst0rm** [comment]
sounds good, ill fix up the style issues and get the pr corrected.

**Review by Ericson2314 — APPROVED** [review]
I corrected a few small things, and this now has my final approval

**Pull request merged**
